From 441bfcc26b6510a9e8f663caf2b157c0dc054cc4 Mon Sep 17 00:00:00 2001
From: Pardha Kesapragada <pkesapragada@habana.ai>
Date: Mon, 28 Sep 2020 11:37:20 +0300
Subject: [PATCH] [SW-21029] ppboot: goya2_arc: move data/bss/stack sections to dccm

In Goya2, PPBoot executes from FW SRAM that is "read-only"
for ARCs. Hence Data/BSS and Stack section updates would not
reflect right values. So move them to DCCM. Since DCCM was
also used by ARCBoot ROM for its internal stack/bss, PPboot
starts using from the second half of DCCM (lower 8KB)

Change-Id: I6ca4efddb5b209a6f4cb6babff8669a8896bd8f2
Signed-off-by: Pardha Kesapragada <pkesapragada@habana.ai>
---

diff --git a/goya2_arc/common.h b/goya2_arc/common.h
index 8caef64..49e56cb 100644
--- a/goya2_arc/common.h
+++ b/goya2_arc/common.h
@@ -17,6 +17,11 @@
 #define SP_SRAM_PREBOOT_OFFSET	0x1000
 #define PPBOOT_TEXT_BASE	(FW_SRAM_BASE + 0xC)
 #define PREBOOT_TEXT_BASE	(SP_SRAM_BASE + SP_SRAM_PREBOOT_OFFSET + 0x8)
+/* In order not to conflict with ARC BOOTROM's usage of DCCM,
+ * Let PPBoot start from the offset of 8KB, for using DCCM
+ * for its own BSS and Stack.
+ */
+#define PPBOOT_DCCM_BASE	0x80002000
 
 #define FUSE_WORDS_PER_BANK	0x40
 #define I2C_SLAVE_ADDRESS	0x6A
diff --git a/goya2_arc/ppboot.lds.S b/goya2_arc/ppboot.lds.S
index 8db0e88..b44b09f 100644
--- a/goya2_arc/ppboot.lds.S
+++ b/goya2_arc/ppboot.lds.S
@@ -7,9 +7,20 @@
 #include "common.h"
 
 #ifdef __arc__
-MEMORY { .sram : ORIGIN = PPBOOT_TEXT_BASE, LENGTH = PPBOOT_SIZE }
 OUTPUT_FORMAT("elf32-littlearc", "elf32-bigarc", "elf32-littlearc")
 OUTPUT_ARCH(arc)
+
+MEMORY
+{
+	.sram 	: ORIGIN = PPBOOT_TEXT_BASE, LENGTH = PPBOOT_SIZE
+	DCCM	: ORIGIN = PPBOOT_DCCM_BASE, LENGTH = 8K
+}
+
+REGION_ALIAS("REGION_DATA",	DCCM)
+REGION_ALIAS("REGION_SDATA",	DCCM)
+REGION_ALIAS("REGION_BSS",	DCCM)
+REGION_ALIAS("REGION_STACK",	DCCM)
+
 #endif
 
 ENTRY(_start)
@@ -18,15 +29,43 @@
 	.text : {
 		. = ALIGN(8);
 		ppboot/crt0_ppboot.o (.text*)
-		* (.text*)
+		*(.text .text.*)
+		_etext = . ;
 	} >.sram
+
+	.rodata : AT ( LOADADDR(.text) + SIZEOF(.text) ) {
+		_rodata = .;
+		*(.rodata .rodata.*)
+		_erodata = . ;
+        } >.sram
+
+	.data : AT ( LOADADDR(.rodata) + SIZEOF(.rodata) ) {
+		. = ALIGN(8);
+		_data = .;
+		*(.data .data.*)
+		_edata = .;
+        } > REGION_DATA
+
+	.sdata : AT ( LOADADDR(.data) + SIZEOF(.data) ) {
+		. = ALIGN(8);
+		_sdata = .;
+		*(.sdata .sdata.* .gnu.linkonce.s.*)
+		_esdata = . ;
+        } > REGION_SDATA
+
+	.bss : {
+		. = ALIGN(8);
+		__bss_start = . ;
+		*(.bss*)
+		__bss_end = . ;
+	} > REGION_BSS
 
 	.stack : {
 		. = ALIGN(8);
 		__stack = . ;
 		. = . + 512 ;
 		__stack_top = . ;
-	} > .sram
+	} > REGION_STACK
 
 	.note.gnu.build-id : { *(.note.gnu.build-id) } > .sram
 }
